pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/service/auth/FENCEAuthenticationService.java

```diff
@@ -159,108 +176,173 @@ public Response getFENCEProfile(String callback_url, Map<String, String> authReq
+        // Also, track the roles that are assigned to the user and in the project_access_set
+        Set<String> rolesAssigned = new HashSet<>();
+        for (Role role : current_user.getRoles()) {
+            if (!project_access_set.contains(role.getName())
+                    && !role.getName().startsWith("MANUAL_")
+                    && !role.getName().equals(fence_open_access_role_name)
+                    && !role.getName().equals("PIC-SURE Top Admin")
+                    && !role.getName().equals("Admin")) {
+                rolesToRemove.add(role);
+            }
 
-        if (current_user.getRoles() != null && (current_user.getRoles().size() > 0 || openAccessIdpValues.contains(idp))) {
+            if (project_access_set.contains(role.getName())) {
+                rolesAssigned.add(role.getName());
+            }
+        }
+
+        // Remove roles that are not in the project_access_set
+        if (!rolesToRemove.isEmpty()) {
+            current_user.getRoles().removeAll(rolesToRemove);
+        }
+
+        // Given the set of roles assigned and that set of roles that should be assigned, we can reduce the set of roles from the project_access_set
+        // to only those that are not in the rolesAssigned set
+        if (!rolesAssigned.isEmpty()) {
+            logger.info("getFENCEProfile() roles that are assigned: {}", rolesAssigned);
+            project_access_set.removeAll(rolesAssigned);
+        }
+
+        if (!project_access_set.isEmpty()) {
+            logger.info("getFENCEProfile() roles that should be assigned: {}", project_access_set);
+
+            // Given our reduced list of roles that should be assigned, we can determine which of those roles are not in the database
+            // This also tells use which roles are in the database
+            Set<String> rolesThatExist = roleRepo.getRoleNamesByNames(project_access_set);
+            if (!rolesThatExist.isEmpty()) {
+                // Assign the roles that exist in the database to the user
+                logger.info("getFENCEProfile() assigning roles that exist in the database: {}", rolesThatExist);
+                Set<Role> rolesThatExistSet = roleRepo.getRolesByNames(rolesThatExist);
+                current_user.getRoles().addAll(rolesThatExistSet);
+
+                // Given a set of all access role names that exist in the database we can now determine which do not exist
+                // and create them
+                project_access_set.removeAll(rolesThatExist);
+            } else {
+                logger.info("getFENCEProfile() none of the following roles exist in the database: {}", project_access_set);
+            }
+
+            // Given the set of all access role names that do not exist in the database we can now create them
+            ArrayList<Role> newRoles = new ArrayList<>();
+            for (String access_role_name : project_access_set) {
+                newRoles.add(createRole(access_role_name, current_user));
```

Comment:
why not do a bulk insert?

Reply:
At this point we are creating roles in the `createRole` method. They are bulk inserted to the database. 
createRole method:
```
private Role createRole(String newRoleName, User current_user) {
        // This is a new Role
        Role r = new Role();
        r.setName(newRoleName);
        r.setDescription("FENCE role "+newRoleName);
        // Since this is a new Role, we need to ensure that the
        // corresponding Privilege (with gates) and AccessRule is added.
        r.setPrivileges(addFENCEPrivileges(current_user, r));
        logger.info("getFENCEProfile() Updated user role. Now it includes `{}`", newRoleName);
        return r;
    }
```
On line 299:
`roleRepo.persistAll(newRoles);`